include/boost/unordered/detail/foa/types_constructibility.hpp

```diff
@@ -0,0 +1,150 @@
+// Copyright (C) 2024 Braden Ganetsky
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_UNORDERED_DETAIL_FOA_TYPES_CONSTRUCTIBILITY_HPP
+#define BOOST_UNORDERED_DETAIL_FOA_TYPES_CONSTRUCTIBILITY_HPP
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace boost {
+  namespace unordered {
+    namespace detail {
+      namespace foa {
+        template <class Key, class... Args> struct check_key_type_t
+        {
+          static_assert(std::is_constructible<Key, Args...>::value,
+            "key_type must be constructible from Args");
+        };
+        template <class Key> struct check_key_type_t<Key, Key&&>
+        {
+          static_assert(std::is_constructible<Key, Key&&>::value,
+            "key_type must be move constructible");
+        };
+
+        template <class Mapped, class... Args> struct check_mapped_type_t
+        {
+          static_assert(std::is_constructible<Mapped, Args...>::value,
+            "mapped_type must be constructible from Args");
+        };
+        template <class Mapped> struct check_mapped_type_t<Mapped, Mapped&&>
+        {
+          static_assert(std::is_constructible<Mapped, Mapped&&>::value,
+            "mapped_type must be move constructible");
+        };
+
```

Comment:
It'd be nice to add a `check_mapped_type_t<Mapped>`  specialization (default construction) to cover this:
```cpp
#include <boost/unordered/unordered_flat_map.hpp>

struct X
{
  X(int){}
};

std::size_t hash_value(const X&){return 0;}
bool operator==(const X&,const X&){return true;}

int main()
{
  boost::unordered_flat_map<int,X> m;
  m[0]=X(0);
}
```

Reply:
Would you recommend a similar specialization for `check_key_type_t<Key>` as well? I'd prefer to add both.